### PR TITLE
Fixed issue with using `Self` ctor in typedefs

### PR DESCRIPTION
**Pull request by alexreg**
Added two regression tests.

This is definitely suitable for a beta backport.

CC @Centril

----

Includes regression tests labelled with issues #56199, #56835, #56611

### CODE REVIEW

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/56225/commits/c39a6422fc5e94b9bb8089d06197051b7ad6ff11 is missing

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -5156,26 +5156,51 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+
+                match adt_def {
+                    Some(adt_def) if adt_def.has_ctor() => {
+                        let variant = adt_def.non_enum_variant();
+                        new_def = Def::StructCtor(variant.did, variant.ctor_kind);
+                        (variant.did, tcx.type_of(variant.did))
+                    }
+                    _ => {
+                        let mut err = tcx.sess.struct_span_err(span,
+                            "the `Self` constructor can only be used with tuple structs");
```
Comment (petrochenkov) [inline src/librustc_typeck/check/mod.rs]: "tuple or unit structs"

**Comment by petrochenkov** [comment]
r=me with error message wording updated (https://github.com/rust-lang/rust/pull/56850#discussion_r241952569)

**Comment by alexreg** [comment]
@petrochenkov LTGM! :-)

**Comment by petrochenkov** [comment]
@alexreg 
Could you squash commits into one if you have time?

**Pull request merged**